bciers/apps/reporting/src/app/components/activities/ActivityForm.tsx

```diff
@@ -0,0 +1,200 @@
+    isChecked: true,
+    elements: [
+      { type: "Page", title: "Review3 ..." },
+      { type: "Page", title: "Other 3.." },
+    ],
+  },
+  { type: "Page", title: "New entrant information", isChecked: true },
+  { type: "Page", title: "Operation emission summary with a long title" },
+];
+
+interface Props {
+  activityData: {
+    activityId: number;
+    sourceTypeMap: { [key: number]: string };
+  };
+  reportDate: string;
+  uiSchema: any;
+  defaultEmptySourceTypeState: any;
+}
+
+// 🧩 Main component
+export default function ActivityForm({
+  activityData,
+  reportDate,
+  uiSchema,
+  defaultEmptySourceTypeState,
+}: Readonly<Props>) {
+  // 🐜 To display errors
+  const [errorList, setErrorList] = useState([] as any[]);
+  // 🌀 Loading state for the Submit button
+  const [isLoading, setIsLoading] = useState(false);
+  // ✅ Success state for for the Submit button
+  const [isSuccess, setIsSuccess] = useState(false);
+  const [formState, setFormState] = useState({} as any);
+  const [jsonSchema, setJsonSchema] = useState({});
+
+  const { activityId, sourceTypeMap } = activityData;
+
+  // Set useEffect dependency set from checked sourceTypes
+  const dependencyArray: string[] = [];
+  for (const value of Object.values(sourceTypeMap)) {
+    dependencyArray.push(
+      formState?.[`${value}`] ? formState?.[`${value}`] : null,
+    );
+  }
+
+  useEffect(() => {
+    let isFetching = true;
+    const fetchSchemaData = async (
+      selectedSourceTypes: string,
+      selectedKeys: number[],
+    ) => {
+      // fetch data from server
+      const schemaData = await actionHandler(
+        `reporting/build-form-schema?activity=${activityId}&report_date=${reportDate}${selectedSourceTypes}`,
+        "GET",
+        "",
+      );
+      setJsonSchema(JSON.parse(schemaData).schema);
+      const sourceTypeFormData = (formState?.sourceTypes as any) || {};
+      // Add an empty sourceType for each selected Source Type (show first item by default)
+      selectedKeys.forEach((k: number) => {
+        if (!formState?.sourceTypes[`${sourceTypeMap[k]}`])
+          sourceTypeFormData[`${sourceTypeMap[k]}`] =
+            defaultEmptySourceTypeState;
+      });
+
+      if (isFetching)
+        setFormState({ ...formState, sourceTypes: sourceTypeFormData });
+    };
+    let selectedSourceTypes = "";
+    const selectedKeys = [];
+    for (const [key, value] of Object.entries(sourceTypeMap)) {
+      if (formState?.[`${value}`]) {
+        selectedSourceTypes = selectedSourceTypes + `&source_types[]=${key}`;
+        selectedKeys.push(Number(key));
+      }
+    }
+
+    fetchSchemaData(selectedSourceTypes, selectedKeys);
+    return () => {
+      isFetching = false;
+    };
+  }, dependencyArray);
+
+  const customFormats = {
+    // Add any needed custom formats here like the example below
+    // phone: /\(?\d{3}\)?[\s-]?\d{3}[\s-]?\d{4}$/,
+  };
+
+  const validator = customizeValidator({ customFormats });
```

Comment:
do we need these?

Reply:
we might? I could take it out until we need it. I thought I'd leave it in there as an example for anyone to add it there if needed. But there are other examples in the code I suppose